# Modeling Modlule

The modeling module provides a single Python API for data scientists to take advantage of the capabilities of ArcGIS as part of demographic geographic data science workflows.

This first cell is largely a bunch of bubblegum and duct tape tying all the things together so I can use this notebook for prototyping and testing.

In [1]:
import importlib
import os
from pathlib import Path
import re
import sys

from dotenv import load_dotenv, find_dotenv
import pandas as pd

# load the "autoreload" extension so that code can change, & always reload modules so that as you change code in src, it gets loaded
%load_ext autoreload
%autoreload 2

# load environment variables from .env
load_dotenv(find_dotenv())

dir_src = Path.cwd().parent.parent.parent/'geosaurus'/'src'

sys.path.insert(0, str(dir_src))

Next, this is where things start looking more like they should. This is where we import the `GIS` object so we can connect to ArcGIS Online and ArcGIS Enterprise environments. We also import the `Country` object and the `get_countries` method from the new `modeling` module.

In [2]:
from arcgis.gis import GIS
from arcgis.modeling import get_countries, Country

## Working Locally

If working in an environment with ArcGIS Pro with Business Analyst and local data, we can introspectively interrogate the local environment to see what is available to work with. As you can see, on my machine, I have a couple of countres with a couple of years worth of data.

In [3]:
get_countries()

iso2 iso3   country_name  vintage     country_id        data_source_id
0   CA  CAN         Canada     2019  CAN_ESRI_2019  LOCAL;;CAN_ESRI_2019
1   US  USA  United States     2019  USA_ESRI_2019  LOCAL;;USA_ESRI_2019
2   US  USA  United States     2020  USA_ESRI_2020  LOCAL;;USA_ESRI_2020

Next, we can instantiate a `Country` object, which will be used for most of the rest of the workflow using the `iso3` identifier. When working with local data, a year can also be optionally specified to work with a specific vintage of data. If not specified, the default is the most current year.

In [4]:
can = Country('CAN')

can

<modeling.Country - CAN (local 2019)>

Next, is is extremely useful to be able to discover the heirarchial geographic levels in a country.

In [5]:
can.levels

geo_name            geo_alias col_id col_name  \
0   disseminationareas   DisseminationAreas     ID     NAME   
1         censustracts         CensusTracts     ID     NAME   
2                 fsas                 FSAs     ID     NAME   
3   censussubdivisions   CensusSubdivisions     ID     NAME   
4                 feds                 FEDs     ID     NAME   
5               cmacas               CMACAs     ID     NAME   
6      censusdivisions      CensusDivisions     ID     NAME   
7  provinceterritories  ProvinceTerritories     ID     NAME   
8              country              Country     ID     NAME   

                                  feature_class_path  
0  D:\arcgis\ba_data\can_2019\Data\Demographic Da...  
1  D:\arcgis\ba_data\can_2019\Data\Demographic Da...  
2  D:\arcgis\ba_data\can_2019\Data\Demographic Da...  
3  D:\arcgis\ba_data\can_2019\Data\Demographic Da...  
4  D:\arcgis\ba_data\can_2019\Data\Demographic Da...  
5  D:\arcgis\ba_data\can_2019\Data\Demographic Da...  
6  D:\arcgis\ba_data\can_2019\Data\Demographic Da...  
7  D:\arcgis\ba_data\can_2019\Data\Demographic Da...  
8  D:\arcgis\ba_data\can_2019\Data\Demographic Da...

Also, access to discover the enrichment variables along with all the ways to reference these variables is extremely useful.

In [6]:
can_ev = can.enrich_variables

can_ev

name                            alias        data_collection  \
0     A16AITFNAT  2016 First Nations Single Ident     AboriginalIdentity   
1      A16AITIDT         2016 Aboriginal Identity     AboriginalIdentity   
2      A16AITIDX       2016 Aboriginal Identities     AboriginalIdentity   
3     A16AITINUK        2016 Inuk Single Identity     AboriginalIdentity   
4     A16AITMETI       2016 Metis Single Identity     AboriginalIdentity   
...          ...                              ...                    ...   
4332   ECYVISKOR                  2019 VM: Korean  VisibleMinorityStatus   
4333  ECYVISJAPA                2019 VM: Japanese  VisibleMinorityStatus   
4334   ECYVISOVM            2019 VM: All Other VM  VisibleMinorityStatus   
4335   ECYVISMVM             2019 VM: Multiple VM  VisibleMinorityStatus   
4336   ECYVISNVM  2019 VM: Not a Visible Minority  VisibleMinorityStatus   

                           enrich_name                 enrich_field_name  
0        AboriginalIdentity.A16AITFNAT     AboriginalIdentity_A16AITFNAT  
1         AboriginalIdentity.A16AITIDT      AboriginalIdentity_A16AITIDT  
2         AboriginalIdentity.A16AITIDX      AboriginalIdentity_A16AITIDX  
3        AboriginalIdentity.A16AITINUK     AboriginalIdentity_A16AITINUK  
4        AboriginalIdentity.A16AITMETI     AboriginalIdentity_A16AITMETI  
...                                ...                               ...  
4332   VisibleMinorityStatus.ECYVISKOR   VisibleMinorityStatus_ECYVISKOR  
4333  VisibleMinorityStatus.ECYVISJAPA  VisibleMinorityStatus_ECYVISJAPA  
4334   VisibleMinorityStatus.ECYVISOVM   VisibleMinorityStatus_ECYVISOVM  
4335   VisibleMinorityStatus.ECYVISMVM   VisibleMinorityStatus_ECYVISMVM  
4336   VisibleMinorityStatus.ECYVISNVM   VisibleMinorityStatus_ECYVISNVM  

[4337 rows x 5 columns]

For instance, if we are only interested in key variables for the current observed year (2019), we can reduce our subset to just these variables. Typically, this is a very useful way to get started and see what type of results a preliminary exploratory modeling effort can discover.

In [7]:
can_ev[
    (can_ev.data_collection.str.lower().str.contains('key'))  # get the key variables
    & (can_ev.name.str.startswith('ECY'))                     # just current year (2019) variables
].reset_index(drop=True)

name                       alias data_collection  \
0    ECYDAYPOP  2019 Daytime Pop Total Pop     KeyCanFacts   
1   ECYWORKPOP    2019 Daytime Pop at Work     KeyCanFacts   
2   ECYHOMEPOP    2019 Daytime Pop at Home     KeyCanFacts   
3    ECYPTAPOP       2019 Total Population     KeyCanFacts   
4     ECYCFSCF  2019 Total Census Families     KeyCanFacts   
5    ECYHSZHHD    2019 HH Size - Total HHs     KeyCanFacts   
6    ECYHNIAVG   2019 HH Inc Average Curr$     KeyCanFacts   
7    ECYHNIMED    2019 HH Inc Median Curr$     KeyCanFacts   
8    ECYTENHHD     2019 Tenure - Total HHs     KeyCanFacts   
9    ECYTENOWN         2019 Tenure - Owned     KeyCanFacts   
10  ECYTENRENT        2019 Tenure - Rented     KeyCanFacts   
11  ECYTENBAND  2019 Tenure - Band Housing     KeyCanFacts   
12    ECYACTPR     2019 Participation Rate     KeyCanFacts   

               enrich_name       enrich_field_name  
0    KeyCanFacts.ECYDAYPOP   KeyCanFacts_ECYDAYPOP  
1   KeyCanFacts.ECYWORKPOP  KeyCanFacts_ECYWORKPOP  
2   KeyCanFacts.ECYHOMEPOP  KeyCanFacts_ECYHOMEPOP  
3    KeyCanFacts.ECYPTAPOP   KeyCanFacts_ECYPTAPOP  
4     KeyCanFacts.ECYCFSCF    KeyCanFacts_ECYCFSCF  
5    KeyCanFacts.ECYHSZHHD   KeyCanFacts_ECYHSZHHD  
6    KeyCanFacts.ECYHNIAVG   KeyCanFacts_ECYHNIAVG  
7    KeyCanFacts.ECYHNIMED   KeyCanFacts_ECYHNIMED  
8    KeyCanFacts.ECYTENHHD   KeyCanFacts_ECYTENHHD  
9    KeyCanFacts.ECYTENOWN   KeyCanFacts_ECYTENOWN  
10  KeyCanFacts.ECYTENRENT  KeyCanFacts_ECYTENRENT  
11  KeyCanFacts.ECYTENBAND  KeyCanFacts_ECYTENBAND  
12    KeyCanFacts.ECYACTPR    KeyCanFacts_ECYACTPR

## Using ArcGIS Online

While it is common to work with a primary country locally, for quite a few large corporations, once they have a successful model running for their primary country, typically the United States, then they want to start experimenting with their international markets. The `modeling` module provides a single interface to explore these hypotheticals without having to install the data locally or learn a new API or interface.

Using the `GIS` object instance connected to ArcGIS Online, we can investigate what countries are available.

In [8]:
agol = GIS(
    os.getenv('ESRI_GIS_URL'),
    username=os.getenv('ESRI_GIS_USERNAME'),
    password=os.getenv('ESRI_GIS_PASSWORD')
)

cntry_df = get_countries(agol)

cntry_df

iso2 iso3 country_name    country_id                           alt_name  \
0     AL  ALB      Albania  ALB_MBR_2019                            ALBANIA   
1     DZ  DZA      Algeria  DZA_MBR_2019                            ALGERIA   
2     AD  AND      Andorra  AND_MBR_2019                            ANDORRA   
3     AO  AGO       Angola  AGO_MBR_2019                             ANGOLA   
4     AR  ARG    Argentina  ARG_MBR_2020                          ARGENTINA   
..   ...  ...          ...           ...                                ...   
131   UY  URY      Uruguay  URY_MBR_2020                            URUGUAY   
132   UZ  UZB   Uzbekistan  UZB_MBR_2020                         UZBEKISTAN   
133   VE  VEN    Venezuela  VEN_MBR_2020  VENEZUELA, BOLIVARIAN REPUBLIC OF   
134   VN  VNM      Vietnam  VNM_MBR_2020                           VIET NAM   
135   ZM  ZMB       Zambia  ZMB_MBR_2019                             ZAMBIA   

         continent  
0           Europe  
1           Africa  
2           Europe  
3           Africa  
4    South America  
..             ...  
131  South America  
132           Asia  
133  South America  
134           Asia  
135         Africa  

[136 rows x 6 columns]

With 136 countires available, rather than scanning the entire dataframe, we can simply see if `GBR` the ISO3 code for Great Britan is available.

In [9]:
cntry_df.iso3.str.contains('GBR').any()

True

In [10]:
cntry_df[cntry_df.iso3.str.contains('GBR')]

iso2 iso3    country_name    country_id        alt_name continent
129   GB  GBR  United Kingdom  GBR_MBR_2019  UNITED KINGDOM    Europe

Since it is, we can create a `Country` object instance just like before, and find the key facts for Great Britian. Obviously each country is going to have a different set of key facts based on what is available. Also, there are a few more columns displaying simply because there are a few more columns available from the REST endpoint than are easily available through local introspection. This, though, is something we are working on to expose locally as well.

In [11]:
gbr = Country('GBR', agol)

gbr_ev = gbr.enrich_variables

gbr_ev[
    (gbr_ev.data_collection.str.lower().str.contains('keyfacts'))
    & (gbr_ev.name.str.endswith('CY'))
].reset_index(drop=True)

name                                              alias  \
0    TOTPOP_CY                              2019 Total Population   
1   POPDENS_CY               2019 Population Density (per sq. km)   
2    POPPRM_CY                           2019 Population Per Mill   
3     MALES_CY                         2019 Total Male Population   
4   FEMALES_CY                       2019 Total Female Population   
5     TOTHH_CY                              2019 Total Households   
6   AVGHHSZ_CY                        2019 Average Household Size   
7    PAGE01_CY                     2019 Total Population Age 0-14   
8    PAGE02_CY                    2019 Total Population Age 15-29   
9    PAGE03_CY                    2019 Total Population Age 30-44   
10   PAGE04_CY                    2019 Total Population Age 45-59   
11   PAGE05_CY                      2019 Total Population Age 60+   
12   MAGE01_CY                      2019 Male Population Age 0-14   
13   MAGE02_CY                     2019 Male Population Age 15-29   
14   MAGE03_CY                     2019 Male Population Age 30-44   
15   MAGE04_CY                     2019 Male Population Age 45-59   
16   MAGE05_CY                       2019 Male Population Age 60+   
17   FAGE01_CY                    2019 Female Population Age 0-14   
18   FAGE02_CY                   2019 Female Population Age 15-29   
19   FAGE03_CY                   2019 Female Population Age 30-44   
20   FAGE04_CY                   2019 Female Population Age 45-59   
21   FAGE05_CY                     2019 Female Population Age 60+   
22   HINC01_CY             2019 HHs: 1st Quintile (Below £20,808)   
23   HINC02_CY        2019 HHs: 2nd Quintile (£20,808 to £30,641)   
24   HINC03_CY        2019 HHs: 3rd Quintile (£30,642 to £41,363)   
25   HINC04_CY        2019 HHs: 4th Quintile (£41,364 to £59,350)   
26   HINC05_CY         2019 HHs: 5th Quintile (£59,351 and above)   
27  HTYP01A_CY                            2019 HHs: Single Person   
28  HTYP02A_CY   2019 HHs: Married Couple with Dependent Children   
29  HTYP03A_CY          2019 HHs: Married Couple with No Children   
30  HTYP04A_CY  2019 HHs: Cohabiting Couple w/ Dependent Children   
31  HTYP05A_CY       2019 HHs: Cohabiting Couple with No Children   
32  HTYP06A_CY    2019 HHs: Single Parent with Dependent Children   
33  HTYP07A_CY             2019 HHs: Single Parent w/ No Children   
34  HTYP08A_CY                   2019 HHs: Multi-Person Household   
35   MRST01_CY                        2019 Marital Status: Single   
36   MRST02_CY                       2019 Marital Status: Married   
37   MRST03_CY                      2019 Marital Status: Divorced   
38   MRST04_CY                       2019 Marital Status: Widowed   
39   EDUC01_CY                2019 Pop 16+/Edu: No Qualifications   
40   EDUC02_CY           2019 Pop 16+/Edu: Level 1 Qualifications   
41   EDUC03_CY           2019 Pop 16+/Edu: Level 2 Qualifications   
42   EDUC04_CY           2019 Pop 16+/Edu: Level 3 Qualifications   
43   EDUC05_CY   2019 Pop 16+/Edu: Level 4 Qualifications & Above   
44   EDUC06_CY               2019 Pop 16+/Edu: Full Time Students   
45    UNEMP_CY                         2018 Unemployed Population   
46       PP_CY                       2019 Purchasing Power: Total   
47    PPPRM_CY                    2019 Purchasing Power: Per Mill   
48     PPPC_CY                  2019 Purchasing Power: Per Capita   
49    PPIDX_CY                       2019 Purchasing Power: Index   

   data_collection          enrich_name    enrich_field_name  \
0         KeyFacts   KeyFacts.TOTPOP_CY   KeyFacts_TOTPOP_CY   
1         KeyFacts  KeyFacts.POPDENS_CY  KeyFacts_POPDENS_CY   
2         KeyFacts   KeyFacts.POPPRM_CY   KeyFacts_POPPRM_CY   
3         KeyFacts    KeyFacts.MALES_CY    KeyFacts_MALES_CY   
4         KeyFacts  KeyFacts.FEMALES_CY  KeyFacts_FEMALES_CY   
5         KeyFacts    KeyFacts.TOTHH_CY    KeyFacts_TOTHH_CY   
6         KeyFacts  KeyFacts.AVGHHSZ_CY  K

## ArcGIS Enterprise (Busines Analyst Server)

Within an organization, if more than just a few data scientists are working with the data, it begins to make sense to migrate local analysis workflows to reference Business Analyst Server as part of an ArcGIS Enterprise installation instead of using ArcGIS Pro with Business Analyst or ArcGIS Online. ArcGIS Pro with Busines Analyst simply does not scale out very well technically or fiscally. ArcGIS Online simply does not scale fiscally. Every time enrichment is run, it costs credits. Thus, it makes sense to quickly be able to move to Business Analyst Server to scale out. The modeling module makes this as easy as simply passing in a GIS object instance connected to an ArcGIS Enterprise instance with Business Analyst Server.

In [12]:
prtl = GIS(
    os.getenv('ESRI_PORTAL_URL'),
    username=os.getenv('ESRI_PORTAL_USERNAME'),
    password=os.getenv('ESRI_PORTAL_PASSWORD')
)

get_countries(prtl)

iso2 iso3   country_name     country_id       alt_name      continent
0   US  USA  United States  USA_ESRI_2020  UNITED STATES  North America

In [13]:
usa = Country('USA', prtl)

usa

<modeling.Country - USA (GIS at https://geoai-ent.bd.esri.com/portal/ logged in as jmccune)>

In [14]:
usa_ev = usa.enrich_variables

usa_ev

name                                              alias  \
0      AGE0_CY                             2020 Population Age <1   
1      AGE1_CY                              2020 Population Age 1   
2      AGE2_CY                              2020 Population Age 2   
3      AGE3_CY                              2020 Population Age 3   
4      AGE4_CY                              2020 Population Age 4   
..         ...                                                ...   
37  MOEMEDYRMV  2018 Median Year Householder Moved In MOE (ACS...   
38  RELMEDYRMV  2018 Median Year Householder Moved In REL (ACS...   
39    ACSOWNER                   2018 Owner Households (ACS 5-Yr)   
40    MOEOWNER               2018 Owner Households MOE (ACS 5-Yr)   
41    RELOWNER               2018 Owner Households REL (ACS 5-Yr)   

    data_collection              enrich_name         enrich_field_name  \
0   1yearincrements  1yearincrements.AGE0_CY  F1yearincrements_AGE0_CY   
1   1yearincrements  1yearincrements.AGE1_CY  F1yearincrements_AGE1_CY   
2   1yearincrements  1yearincrements.AGE2_CY  F1yearincrements_AGE2_CY   
3   1yearincrements  1yearincrements.AGE3_CY  F1yearincrements_AGE3_CY   
4   1yearincrements  1yearincrements.AGE4_CY  F1yearincrements_AGE4_CY   
..              ...                      ...                       ...   
37      yearmovedin   yearmovedin.MOEMEDYRMV    yearmovedin_MOEMEDYRMV   
38      yearmovedin   yearmovedin.RELMEDYRMV    yearmovedin_RELMEDYRMV   
39      yearmovedin     yearmovedin.ACSOWNER      yearmovedin_ACSOWNER   
40      yearmovedin     yearmovedin.MOEOWNER      yearmovedin_MOEOWNER   
41      yearmovedin     yearmovedin.RELOWNER      yearmovedin_RELOWNER   

                                          description    vintage  units  
0                 2020 Total Population Age <1 (Esri)       2020  count  
1                  2020 Total Population Age 1 (Esri)       2020  count  
2                  2020 Total Population Age 2 (Esri)       2020  count  
3                  2020 Total Population Age 3 (Esri)       2020  count  
4                  2020 Total Population Age 4 (Esri)       2020  count  
..                                                ...        ...    ...  
37  2018 Median Year Householder Moved into Unit M...  2014-2018  count  
38  2018 Median Year Householder Moved into Unit R...  2014-2018  count  
39                   2018 Owner Households (ACS 5-Yr)  2014-2018  count  
40               2018 Owner Households MOE (ACS 5-Yr)  2014-2018  count  
41               2018 Owner Households REL (ACS 5-Yr)  2014-2018  count  

[17474 rows x 8 columns]

In [15]:
usa_ev[
    (usa_ev.data_collection.str.lower().str.contains('key'))
    & (usa_ev.name.str.endswith('CY'))
].reset_index(drop=True)

name                              alias data_collection  \
0    TOTPOP_CY              2020 Total Population      KeyUSFacts   
1     GQPOP_CY     2020 Group Quarters Population      KeyUSFacts   
2   DIVINDX_CY               2020 Diversity Index      KeyUSFacts   
3     TOTHH_CY              2020 Total Households      KeyUSFacts   
4   AVGHHSZ_CY        2020 Average Household Size      KeyUSFacts   
5   MEDHINC_CY       2020 Median Household Income      KeyUSFacts   
6   AVGHINC_CY      2020 Average Household Income      KeyUSFacts   
7       PCI_CY             2020 Per Capita Income      KeyUSFacts   
8     TOTHU_CY           2020 Total Housing Units      KeyUSFacts   
9     OWNER_CY            2020 Owner Occupied HUs      KeyUSFacts   
10   RENTER_CY           2020 Renter Occupied HUs      KeyUSFacts   
11   VACANT_CY          2020 Vacant Housing Units      KeyUSFacts   
12   MEDVAL_CY             2020 Median Home Value      KeyUSFacts   
13   AVGVAL_CY            2020 Average Home Value      KeyUSFacts   
14  POPGRW10CY  2010-2020 Growth Rate: Population      KeyUSFacts   
15   HHGRW10CY  2010-2020 Growth Rate: Households      KeyUSFacts   
16  FAMGRW10CY    2010-2020 Growth Rate: Families      KeyUSFacts   
17     DPOP_CY      2020 Total Daytime Population      KeyUSFacts   
18  DPOPWRK_CY          2020 Daytime Pop: Workers      KeyUSFacts   
19  DPOPRES_CY        2020 Daytime Pop: Residents      KeyUSFacts   

              enrich_name      enrich_field_name  \
0    KeyUSFacts.TOTPOP_CY   KeyUSFacts_TOTPOP_CY   
1     KeyUSFacts.GQPOP_CY    KeyUSFacts_GQPOP_CY   
2   KeyUSFacts.DIVINDX_CY  KeyUSFacts_DIVINDX_CY   
3     KeyUSFacts.TOTHH_CY    KeyUSFacts_TOTHH_CY   
4   KeyUSFacts.AVGHHSZ_CY  KeyUSFacts_AVGHHSZ_CY   
5   KeyUSFacts.MEDHINC_CY  KeyUSFacts_MEDHINC_CY   
6   KeyUSFacts.AVGHINC_CY  KeyUSFacts_AVGHINC_CY   
7       KeyUSFacts.PCI_CY      KeyUSFacts_PCI_CY   
8     KeyUSFacts.TOTHU_CY    KeyUSFacts_TOTHU_CY   
9     KeyUSFacts.OWNER_CY    KeyUSFacts_OWNER_CY   
10   KeyUSFacts.RENTER_CY   KeyUSFacts_RENTER_CY   
11   KeyUSFacts.VACANT_CY   KeyUSFacts_VACANT_CY   
12   KeyUSFacts.MEDVAL_CY   KeyUSFacts_MEDVAL_CY   
13   KeyUSFacts.AVGVAL_CY   KeyUSFacts_AVGVAL_CY   
14  KeyUSFacts.POPGRW10CY  KeyUSFacts_POPGRW10CY   
15   KeyUSFacts.HHGRW10CY   KeyUSFacts_HHGRW10CY   
16  KeyUSFacts.FAMGRW10CY  KeyUSFacts_FAMGRW10CY   
17     KeyUSFacts.DPOP_CY     KeyUSFacts_DPOP_CY   
18  KeyUSFacts.DPOPWRK_CY  KeyUSFacts_DPOPWRK_CY   
19  KeyUSFacts.DPOPRES_CY  KeyUSFacts_DPOPRES_CY   

                                        description vintage     units  
0                      2020 Total Population (Esri)    2020     count  
1             2020 Group Quarters Population (Esri)    2020     count  
2                       2020 Diversity Index (Esri)    2020     count  
3                      2020 Total Households (Esri)    2020     count  
4                2020 Average Household Size (Esri)    2020     count  
5               2020 Median Household Income (Esri)    2020  currency  
6              2020 Average Household Income (Esri)    2020  currency  
7                     2020 Per Capita Income (Esri)    2020  currency  
8                   2020 Total Housing Units (Esri)    2020     count  
9          2020 Owner Occupied Housing Units (Esri)    2020     count  
10        2020 Renter Occupied Housing Units (Esri)    2020     count  
11                 2020 Vacant Housing Units (Esri)    2020     count  
12                    2020 Median Home Value (Esri)    2020  currency  
13                   2020 Average Home Value (Esri)    2020  currency  
14  2010-2020 Population: Annual Growth Rate (Esri)    2020       pct  
15  2010-2020 Households: Annual Growth Rate (Esri)    2020       pct  
16    2010-2020 Families: Annual Growth Rate (Esri)    2020       pct  
17             2020 Total Daytime Population (Esri)    2020     count  
18          2020 Daytime Population: Workers (Esri)    2020     count  
19        2020 Daytime Popu

## Flexiblity Illustrated

While each of the above examples illustrated different data sources, just to illustrate the ease of moving between these. Here is the same workflow for the United States in ArcGIS Online and with local data.

In [16]:
usa = Country('USA', agol)
usa_ev = usa.enrich_variables
usa_ev[
    (usa_ev.data_collection.str.lower().str.contains('key'))
    & (usa_ev.name.str.endswith('CY'))
].reset_index(drop=True)

name                              alias data_collection  \
0    TOTPOP_CY              2020 Total Population      KeyUSFacts   
1     GQPOP_CY     2020 Group Quarters Population      KeyUSFacts   
2   DIVINDX_CY               2020 Diversity Index      KeyUSFacts   
3     TOTHH_CY              2020 Total Households      KeyUSFacts   
4   AVGHHSZ_CY        2020 Average Household Size      KeyUSFacts   
5   MEDHINC_CY       2020 Median Household Income      KeyUSFacts   
6   AVGHINC_CY      2020 Average Household Income      KeyUSFacts   
7       PCI_CY             2020 Per Capita Income      KeyUSFacts   
8     TOTHU_CY           2020 Total Housing Units      KeyUSFacts   
9     OWNER_CY            2020 Owner Occupied HUs      KeyUSFacts   
10   RENTER_CY           2020 Renter Occupied HUs      KeyUSFacts   
11   VACANT_CY          2020 Vacant Housing Units      KeyUSFacts   
12   MEDVAL_CY             2020 Median Home Value      KeyUSFacts   
13   AVGVAL_CY            2020 Average Home Value      KeyUSFacts   
14  POPGRW10CY  2010-2020 Growth Rate: Population      KeyUSFacts   
15   HHGRW10CY  2010-2020 Growth Rate: Households      KeyUSFacts   
16  FAMGRW10CY    2010-2020 Growth Rate: Families      KeyUSFacts   
17     DPOP_CY      2020 Total Daytime Population      KeyUSFacts   
18  DPOPWRK_CY          2020 Daytime Pop: Workers      KeyUSFacts   
19  DPOPRES_CY        2020 Daytime Pop: Residents      KeyUSFacts   

              enrich_name      enrich_field_name  \
0    KeyUSFacts.TOTPOP_CY   KeyUSFacts_TOTPOP_CY   
1     KeyUSFacts.GQPOP_CY    KeyUSFacts_GQPOP_CY   
2   KeyUSFacts.DIVINDX_CY  KeyUSFacts_DIVINDX_CY   
3     KeyUSFacts.TOTHH_CY    KeyUSFacts_TOTHH_CY   
4   KeyUSFacts.AVGHHSZ_CY  KeyUSFacts_AVGHHSZ_CY   
5   KeyUSFacts.MEDHINC_CY  KeyUSFacts_MEDHINC_CY   
6   KeyUSFacts.AVGHINC_CY  KeyUSFacts_AVGHINC_CY   
7       KeyUSFacts.PCI_CY      KeyUSFacts_PCI_CY   
8     KeyUSFacts.TOTHU_CY    KeyUSFacts_TOTHU_CY   
9     KeyUSFacts.OWNER_CY    KeyUSFacts_OWNER_CY   
10   KeyUSFacts.RENTER_CY   KeyUSFacts_RENTER_CY   
11   KeyUSFacts.VACANT_CY   KeyUSFacts_VACANT_CY   
12   KeyUSFacts.MEDVAL_CY   KeyUSFacts_MEDVAL_CY   
13   KeyUSFacts.AVGVAL_CY   KeyUSFacts_AVGVAL_CY   
14  KeyUSFacts.POPGRW10CY  KeyUSFacts_POPGRW10CY   
15   KeyUSFacts.HHGRW10CY   KeyUSFacts_HHGRW10CY   
16  KeyUSFacts.FAMGRW10CY  KeyUSFacts_FAMGRW10CY   
17     KeyUSFacts.DPOP_CY     KeyUSFacts_DPOP_CY   
18  KeyUSFacts.DPOPWRK_CY  KeyUSFacts_DPOPWRK_CY   
19  KeyUSFacts.DPOPRES_CY  KeyUSFacts_DPOPRES_CY   

                                        description vintage     units  
0                      2020 Total Population (Esri)    2020     count  
1             2020 Group Quarters Population (Esri)    2020     count  
2                       2020 Diversity Index (Esri)    2020     count  
3                      2020 Total Households (Esri)    2020     count  
4                2020 Average Household Size (Esri)    2020     count  
5               2020 Median Household Income (Esri)    2020  currency  
6              2020 Average Household Income (Esri)    2020  currency  
7                     2020 Per Capita Income (Esri)    2020  currency  
8                   2020 Total Housing Units (Esri)    2020     count  
9          2020 Owner Occupied Housing Units (Esri)    2020     count  
10        2020 Renter Occupied Housing Units (Esri)    2020     count  
11                 2020 Vacant Housing Units (Esri)    2020     count  
12                    2020 Median Home Value (Esri)    2020  currency  
13                   2020 Average Home Value (Esri)    2020  currency  
14  2010-2020 Population: Annual Growth Rate (Esri)    2020       pct  
15  2010-2020 Households: Annual Growth Rate (Esri)    2020       pct  
16    2010-2020 Families: Annual Growth Rate (Esri)    2020       pct  
17             2020 Total Daytime Population (Esri)    2020     count  
18          2020 Daytime Population: Workers (Esri)    2020     count  
19        2020 Daytime Popu

In [17]:
usa = Country('USA', 'local')
usa_ev = usa.enrich_variables
usa_ev[
    (usa_ev.data_collection.str.lower().str.contains('key'))
    & (usa_ev.name.str.endswith('CY'))
].reset_index(drop=True)

name                              alias data_collection  \
0    TOTPOP_CY              2020 Total Population      KeyUSFacts   
1     GQPOP_CY     2020 Group Quarters Population      KeyUSFacts   
2   DIVINDX_CY               2020 Diversity Index      KeyUSFacts   
3     TOTHH_CY              2020 Total Households      KeyUSFacts   
4   AVGHHSZ_CY        2020 Average Household Size      KeyUSFacts   
5   MEDHINC_CY       2020 Median Household Income      KeyUSFacts   
6   AVGHINC_CY      2020 Average Household Income      KeyUSFacts   
7       PCI_CY             2020 Per Capita Income      KeyUSFacts   
8     TOTHU_CY           2020 Total Housing Units      KeyUSFacts   
9     OWNER_CY            2020 Owner Occupied HUs      KeyUSFacts   
10   RENTER_CY           2020 Renter Occupied HUs      KeyUSFacts   
11   VACANT_CY          2020 Vacant Housing Units      KeyUSFacts   
12   MEDVAL_CY             2020 Median Home Value      KeyUSFacts   
13   AVGVAL_CY            2020 Average Home Value      KeyUSFacts   
14  POPGRW10CY  2010-2020 Growth Rate: Population      KeyUSFacts   
15   HHGRW10CY  2010-2020 Growth Rate: Households      KeyUSFacts   
16  FAMGRW10CY    2010-2020 Growth Rate: Families      KeyUSFacts   
17     DPOP_CY      2020 Total Daytime Population      KeyUSFacts   
18  DPOPWRK_CY          2020 Daytime Pop: Workers      KeyUSFacts   
19  DPOPRES_CY        2020 Daytime Pop: Residents      KeyUSFacts   

              enrich_name      enrich_field_name  
0    KeyUSFacts.TOTPOP_CY   KeyUSFacts_TOTPOP_CY  
1     KeyUSFacts.GQPOP_CY    KeyUSFacts_GQPOP_CY  
2   KeyUSFacts.DIVINDX_CY  KeyUSFacts_DIVINDX_CY  
3     KeyUSFacts.TOTHH_CY    KeyUSFacts_TOTHH_CY  
4   KeyUSFacts.AVGHHSZ_CY  KeyUSFacts_AVGHHSZ_CY  
5   KeyUSFacts.MEDHINC_CY  KeyUSFacts_MEDHINC_CY  
6   KeyUSFacts.AVGHINC_CY  KeyUSFacts_AVGHINC_CY  
7       KeyUSFacts.PCI_CY      KeyUSFacts_PCI_CY  
8     KeyUSFacts.TOTHU_CY    KeyUSFacts_TOTHU_CY  
9     KeyUSFacts.OWNER_CY    KeyUSFacts_OWNER_CY  
10   KeyUSFacts.RENTER_CY   KeyUSFacts_RENTER_CY  
11   KeyUSFacts.VACANT_CY   KeyUSFacts_VACANT_CY  
12   KeyUSFacts.MEDVAL_CY   KeyUSFacts_MEDVAL_CY  
13   KeyUSFacts.AVGVAL_CY   KeyUSFacts_AVGVAL_CY  
14  KeyUSFacts.POPGRW10CY  KeyUSFacts_POPGRW10CY  
15   KeyUSFacts.HHGRW10CY   KeyUSFacts_HHGRW10CY  
16  KeyUSFacts.FAMGRW10CY  KeyUSFacts_FAMGRW10CY  
17     KeyUSFacts.DPOP_CY     KeyUSFacts_DPOP_CY  
18  KeyUSFacts.DPOPWRK_CY  KeyUSFacts_DPOPWRK_CY  
19  KeyUSFacts.DPOPRES_CY  KeyUSFacts_DPOPRES_CY